### PR TITLE
[DeviceASAN] Simplify shadow mapping logic for GPU device

### PR DESCRIPTION
**Pull request by zhaomaosu**
intel/llvm part: https://github.com/intel/llvm/pull/15954

### CODE REVIEW

File: source/loader/layers/sanitizer/asan_shadow.cpp [hunk 1]
```diff
@@ -256,12 +256,9 @@ ur_result_t ShadowMemoryGPU::ReleaseShadow(std::shared_ptr<AllocInfo> AI) {
...
 uptr ShadowMemoryPVC::MemToShadow(uptr Ptr) {
-    if (Ptr & 0xFF00000000000000ULL) { // Device USM
-        return ShadowBegin + 0x80000000000ULL +
-               ((Ptr & 0xFFFFFFFFFFFFULL) >> ASAN_SHADOW_SCALE);
-    } else { // Only consider 47bit VA
-        return ShadowBegin + ((Ptr & 0x7FFFFFFFFFFFULL) >> ASAN_SHADOW_SCALE);
-    }
+    return ShadowBegin + (((((Ptr & 0x8000'0000'0000'0000) >> 16) + Ptr) &
+                           0xf'ffff'ffff'ffff) >>
+                          ASAN_SHADOW_SCALE);
```
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/asan_shadow.cpp:261]: Could we add some comment here to explain the mapping logic?

Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_shadow.cpp:261]: Actually, we already have shadow mapping logic here https://github.com/zhaomaosu/unified-runtime/blob/simplify-shadow-mapping/source/loader/layers/sanitizer/asan_shadow.hpp#L99

Comment (yingcong-wu) [reply source/loader/layers/sanitizer/asan_shadow.cpp:261]: okay

**Review by AllanZyne — APPROVED** [review]

**Comment by AllanZyne** [comment]
Change the mapping logic of DG2 as well?

**Comment by zhaomaosu** [comment]
> Let's change the mapping logic of DG2 as well?

Done

File: source/loader/layers/sanitizer/ur_sanddi.cpp [hunk 2]
```diff
@@ -73,6 +73,12 @@ __urdlllocal ur_result_t UR_APICALL urAdapterGet(
         return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
     }
 
+    // Set env AllocateHostAllocationsInHeapExtendedHost to 0, then host and
+    // shared usm will have same allocation range on GPU device. This will
+    // simplify shadow memory mapping logic on GPU device.
+    setenv("NEOReadDebugKeys", "1", 1);
```
Comment (pbalcer) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:79]: @nrspruit is it a good idea to set these env variables in the layer? If this is really required, wouldn't it be a better idea to make this into an L0 API?
Using a debug key for functionality doesn't seem like a good idea.

Comment (pbalcer) [reply source/loader/layers/sanitizer/ur_sanddi.cpp:79]: I don't feel comfortable merging this patch. We need confirmation from the NEO folks that this makes sense.
@MichalMrozek can you help?

Comment (MichalMrozek) [reply source/loader/layers/sanitizer/ur_sanddi.cpp:79]: No, this is not correct to set this.
Please do not merge it.
If you want to have specific release keys avaialable this needs discussion with Piotr Rozenfeld.

Comment (MichalMrozek) [reply source/loader/layers/sanitizer/ur_sanddi.cpp:79]: AllocateHostAllocationsInHeapExtendedHost is a debug key, debug keys cannot be used in production software. They are meant only for experiments. Driver team may remove or change this debug variable at any point of time without any notifications.

Comment (AllanZyne) [reply source/loader/layers/sanitizer/ur_sanddi.cpp:79]: Ok, I will send this request to driver team. Thanks.

**Comment by zhaomaosu** [comment]
I'll try to find other optimization opportunity. I'm closing this PR.

**Pull request closed without merging**